README.md
# Anybus CompactCom SPI Protocol Analyzer Plugin

[![windows_build](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/windows_build.yml/badge.svg)](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/windows_build.yml)
[![ubuntu_build](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/ubuntu_build.yml/badge.svg)](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/ubuntu_build.yml)
[![macos_build](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/macos_build.yml/badge.svg)](https://github.com/hms-networks/AbccSpiAnalyzer/actions/workflows/macos_build.yml)

## Copyright &copy; 2015-2022 HMS Industrial Networks, Inc.

THE CODE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. HMS DOES NOT
WARRANT THAT THE FUNCTIONS OF THE CODE WILL MEET YOUR REQUIREMENTS, OR
THAT THE OPERATION OF THE CODE WILL BE UNINTERRUPTED OR ERROR-FREE, OR
THAT DEFECTS IN IT CAN BE CORRECTED.

---

## [Table of Contents](#table-of-contents)

1. [Description](#description)
2. [System Requirements](#system-requirements)
3. [Precompiled Releases](#precompiled-releases)
4. [Compile & Install](#compile-&-install)
   * [Initial Steps](#initial-steps)
   * [Windows](#windows)
   * [GNU/Linux](#gnulinux)
   * [macOS](#macos)
5. [Generating Releases](#generating-releases)
6. [Documentation](#documentation)
7. [Changelog](#changelog)
8. [Licenses](#licenses)

---

## [Description](#table-of-contents)

This plugin, in conjunction with the [Saleae Logic][link_saleae] hardware and
software, provides an easy-to-use interface for analyzing bi-directional (full-duplex)
[Anybus CompactCom (ABCC)](https://www.anybus.com/products/embedded-index) SPI
protocol communication.

Each field within an SPI packet is added as a multi-layered bubble-text
within the Logic software. Basic markup is displayed when zoomed-out, while
more detailed information is displayed when zoomed-in.

The decoded results are indexed and searchable within the Logic software; for
even more detail, the user can export the message data or process data to a CSV
file to be imported into Excel (or similar) for further analysis. For instance,
this functionality may reduce the effort required to extract file data
transferred from ABCC object messaging, or to plot a waveform extracted from
process data using either local timestamp information or the network timestamps
(if supported by the network protocol).

![Overview of Plugin][mov_overview]

## [System Requirements](#table-of-contents)

* Saleae Logic Software
  * **Recommended Version**: 1.2.29
  * Other versions will work so long as **Analyzer SDK version 1.1.32** is
    compatible.
  * **NOTE**: At this time, Saleae Logic V2 is not supported.
* Saleae Logic Hardware
  * While any of Saleae Logic hardware offerings is expected to be compatible,
  it is important to consider sample rate limitations of the Logic hardware and
  what SPI data rate the ABCC host-module communication is using. The absolute
  minimum requirement for sufficiently sampling the communication is 2x the
  frequency of the SPI SCLK. The ABCC is able to work over a wide range of
  clock rates where the maximum is specified at 20MHz. So to be able to support
  this would require at a minimum the ability to sample at 40MSamples/second.
  Keep in mind that these analyzers may reduce sample rate depending on how many
  channels on the analyzer are active, so it is imperative that the hardware
  is able to sufficiently sample the SPI bus for the number of channels required.
* PC running:
  * Windows 7 through Windows 10
    * NOTE: At this time Window 11 has compatibility issues with version 1.2.29.
      Version 1.2.40 resolves this compatibility issue but Saleae dropped
      support for the decoded protocols subwindow for this version. An
      alternative solution for using 1.2.29 on Windows 11 would be to use `WSLg`
      and attach the Saleae Logic interface via `usbpid.exe`.
      See [How-To: Connect USB devices](https://docs.microsoft.com/en-us/windows/wsl/connect-usb).
  * Ubuntu 12.04.2+
  * Mac OSX 10.7 Lion+
* USB 2.0 or better
  * Using the Saleae Logic on a dedicated USB controller that matches or exceeds
  the Logic's supported USB version is recommended for ensuring the best
  possible sampling performance.

## [Precompiled Releases](#table-of-contents)

Precompiled plugins and documentation are available here: [Plugin Releases][link_releases]

## [Compile & Install](#table-of-contents)

While tagged releases are provided which contain pre-compiled libraries and
associated documentation, the user may still opt to compile the libraries
for themselves.

### [Initial Steps](#table-of-contents)

After cloning this git repository, the user will need to initialize the
submodules command in order to fetch the repositories containing the Logic
Analyzer SDK. Two copies are used of the same repo, one pointed at the
release/master branch and another pointing at the legacy SDK to support debugging.

Example for cloning the repository and fetching/initializing the submodules:

```bash
git clone https://github.com/hms-networks/AbccSpiAnalyzer.git
cd AbccSpiAnalyzer
git submodule update --init --recursive
```

### [Windows](#table-of-contents)

To compile the project please ensure the **additional** requirements are met:

> DEPENDENCIES: **Visual Studio 2019**, other versions may also work
*(project retargeting may be necessary)*

> NOTE: **Express** or **Community** versions of Visual Studio may require
additional steps to replace occurrences of `#include "afxres.h"` with
`#include "windows.h"` or installation of the Visual C++ MFC package (this can
be done by modifying the Visual Studio installation and enabling the associated
feature).

When compiling the project, please ensure that the correct library is linked
for the Windows OS being used. The VS solutions provided are setup with the
expectation that the host system will be x64. The solutions are also setup for
batch build such that an x64 system can compile all supported build configurations
with one request. To access this, right-click the **Solution 'AbccSpiAnalyzer'**
item in the **Solution Explorer** subwindow to access the context menu and select
**Batch Build...**.

#### Cross-compile for GNU/Linux Operating Systems

Also included in the repository is a simple batch file that can be used to quickly
build both the Windows and GNU/Linux libraries via the use of Windows 10 WSL
(Windows Subsystem for Linux). Cross-compiling for GNU/Linux requires the same
set of dependencies documented below for GNU/Linux.

#### Custom Visual Studio Projects

If there is a need to create a new Visual Studio project, ensure the linker is
configured to include the correct library:

* If using **Win64**, specify `./sdk/release/Analyzer64.lib` in the linker input.
  * The configuration manager has this setup as x64.
* If using **Win32**, specify `./sdk/release/Analyzer.lib` in the linker input.
  * The configuration manager has this setup as Win32.

Once compiled, a file called `AbccSpiAnalyzer.dll` or `AbccSpiAnalyzer64.dll`
in the either `./plugins/Win32/` or `./plugins/Win64` folder, respectively. Copy
this DLL to the user's Saleae Logic software installation in the "Analyzers" folder:

* Example: `C:\Program Files\Saleae LLC\Analyzers\AbccSpiAnalyzer.dll`

The analyzer is now ready to be used and can be found in the same way that
other protocol analyzers are added to the Logic software.

### [GNU/Linux](#table-of-contents)

* To compile simply run the `build_analyzer.py` python script. This script will
  compile the appropriate shared object library based on the host system.

> DEPENDENCIES: **Python**, **G++**

> ADDITIONAL DEPENDENCY: Depending on the system it may also be required to
install `g++-multilib` in order for the python build script to complete without
error. This is so a 64-bit system can cross-compile for 32-bit machines.

Once compiled, a file called `AbccSpiAnalyzer.so` or `AbccSpiAnalyzer64.so`
will reside in the either `./plugins/Linux/` or `./plugins/Linux64` folder,
respectively. Copy this shared object to the user's Saleae Logic software
installation in the "Analyzers" folder.

### [macOS](#table-of-contents)

* To compile simply run the `build_analyzer.py` python script.

> DEPENDENCIES: **Python**, **G++**

Once compiled, a dynamic library called `AbccSpiAnalyzer.dylib` will reside
in the `./plugins/OSX/` folder. Copy this dynamic object to the user's Saleae
Logic software installation in the "Analyzers" folder.

### [Generating Releases](#table-of-contents)

This section is not typically applicable for most users, but is documented here
for completeness. The python script `make_release.py` will generate the ZIP file
used for releases. However to use this without error a user must collect the
compiled library files for each platform. On a Windows 64-bit host with Ubuntu
bash on Windows using WSL (Windows Subsystem for Linux), it is possible to
generate all required library files with the exception to macOS using
`cross_compile.bat`. Obtaining the macOS dynamic library is a separate manual
step, transferring the dylib from the macOS build host to the Windows host.

Windows:

```bash
py.exe -3 -m pip install -r py_requirements.txt
py.exe -3 ./make_release.py
```

macOS/Linux:

```bash
python3 -m pip install -r py_requirements.txt
python3 ./make_release.py
```

## [Documentation](#table-of-contents)

For complete details regarding this plugin's functionality please consult the
[Wiki][link_wiki] found within this repository or [Quick Start Guide][link_qsg]
included in this repository's documentation folder. The Wiki is expected to
contain more up-to-date details on the plugin's functionality in reference to
the master branch whereas the PDF document will typically be updated when a new
tagged release is made and is provided as part of the tagged release's
downloadable assets.

For details on the Logic SDK's API please refer to the Saleae's
[Protocol Analyzer SDK][link_sdk] page.

## [Changelog](#table-of-contents)

Please see [CHANGELOG.md](CHANGELOG.md) provided within this repository for details on the changelog history.

## [Licenses](#table-of-contents)

Please see [LICENSE.md](LICENSE.md) provided within this repository for details on the software licensing.

[mov_overview]: https://github.com/HMSAB/AbccSpiAnalyzer/wiki/overview.gif "Overview of Plugin"
[link_wiki]: https://github.com/HMSAB/AbccSpiAnalyzer/wiki
[link_releases]: https://github.com/HMSAB/AbccSpiAnalyzer/releases
[link_qsg]: doc/AbccSpiAnalyzer_Plugin_Quick_Start_Guide.pdf
[link_sdk]: https://support.saleae.com/saleae-api-and-sdk/protocol-analyzer-sdk
[link_saleae]: https://www.saleae.com/


CHANGELOG.md
# Changelog

## Version 2022.6.10.1

### Changes

* Code refactoring.
* Updated ABP headers and lookup tables.
* Improved support for log files containing BOM and added support for UTF16.
  This means the following formats are acceptable: ASCII, UTF-8, UTF-8-BOM,
  UTF-16-BOM.
* Updated ABP_CIPID_IA_PRODUCT_NAME to support both numeric and character
  display.

### Fixes

* Fixed NC object's instance name lookup logic.
* Fixed ABP_NW_EXCPT_INFO_NONE, the verbose string was incorrect.

---

## Version 2021.6.1.1

### Changes

* Code refactoring
* Added support for ABCC SDK "log file simulation". It is now possible to take
  a standard SDK log file and simulate SPI messaging in the plugin via
  the `AdvancedSettings.xml` file. See plugin documentation for more information.
  NOTE: Some of these new configurable options are also available when using
  "standard simulation" mode. Example log files have been added to the
  `./doc/abcc_log_file_examples/` folder.
* Added support for message segmentation.
* Added support for enumeration of exception information and codes.
* Added support for message header fragmentation. Previously it was not possible
  for the plugin to handle cases where the ABCC message header was split
  between multiple SPI packets.
* Improved support for error response reporting. Error response is now included
  in the exported data and decoded protocols output.

### Fixes

* Resolved static analysis issues.
* Fixed an issue with byte acquisition logic. If the clock transitions from high
  to low after the last bit in the byte has already been sampled but before the
  chip select deasserts, the plugin incorrectly reported this event as an error
  which in turn caused the SPI packet to be incorrectly marked with an error.
* Fixed bug regarding object specific errors. This issue resulted in an
  "Unknown" string entry being reported when it should not have done so.
* Added workaround setting to `AdvancedSettings.xml` to fix an inherent bug in
  the Logic software version 1.2.x. Error frames committed by the logic analyzer
  plugin which span < 8 samples require "expansion" to 8 samples in order for
  the software to properly establish a link between the decoded protocols entry
  and the bubble text.
* Improved CSV character escaping logic when using certain delimiters.
* Minor spelling fixes.

---

## Version 2021.1.5.1

### Changes

* Code refactoring
* Updated lookup tables to support latest ABP additions.
* Changed channel names to more closely match ABCC documentation.

### Fixes

* Resolved static analysis issues.

---

## Version 2019.07.17.1

### Changes

* Updated lookup tables to contain latest ABP header information.
* Migrated from VS2017 to VS2019
* Other minor changes & refactoring

### Fixes

* (Issue #23) Fixed the state logic when the enable-signal toggles and there is
  no clocked data.
* (Issue #24) Fixed the logic behind multi-event markers.
* Fixed build issue for macOS Mojave.

---

## Version 2018.10.10.1

### Changes

* Significant internal code refactoring
* Updated lookup tables to contain latest ABP header information.
* Added support for network specific error codes for Profinet Diagnostic object
* Significant rework of the plugin's simulation was done to improve highlighting
  the functionality of the plugin. The simulation now performs a more complex
  request-response messaging sequence that targets reading file information from
  the plugin's metadata for the compiled DLL. This doubles to provide built-in
  version information into the plugin for operating systems that do not contain
  such metadata for library objects.
* Significant rework of the CSV file exporting options for ABCC message data and
  process data. These exported files now include state information of the network
  and application. The CSV for message data now provides enumerated results
  making it much more human readable. Error event indicators are also now included
  to help point out CRC error, fragmented SPI packets, and retransmitted packets.
* Changed how the "display base" is used in the plugin. Now, only fields that are
  known to be of a 'character' base-type will be permitted to be displayed as
  "ASCII" or "ASCII & HEX". Frames that are not a "CHAR" base-type will be
  displayed in HEX for both of these options. In general, this means the ASCII
  display options will only apply to the "Message Data" field. This should provide
  a more intuitive user-experience in interpreting the information contained in
  the packet as a whole. Both attribute and command message data have been
  updated to follow these rules. For instance, the File System Interface's
  File_Open message data contains a filename that is naturally understood as a
  string which is classified as a 'character' base-type; whereas the "File Size"
  attribute is naturally understood as a 'numeric' data field. For any field,
  where it may be desirable to permit display in ASCII the BaseType::Character
  should be used. Future adjustments may be made to a particular command or
  attribute BaseType.
* Removed network types from settings that are ABCC 30-series only.
* This changelog has been modified to remove the "known issues" section. These
  are likely to remain as limitations of the software for some time and a separate
  document has been added to track these.
* Added option to limit the number of "clocking alerts" to report to the UI.
  This can be accessed from the advanced settings XML file.

### Fixes

* Fixed lookup logic for the "Object" instance name. Previously, instance 0 was
  being reported as "Unknown" instead of "Object".
* Fixed "compact" indexing MOSI tag. The MOSI channels indexed results were
  mistakenly using the MISO_TAG_STR instead of MOSI_TAG_STR.
* Fixed issue with last packet not being committed. The last packet in a capture

build_analyzer.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import os
import glob
import platform
import subprocess

"""
Builds the plugin for the current platform and when appropriate for both x86
and x64 architectures.
"""

# If MSBuild is not included in system path, full path can be specified here
MSBUILD_EXE = "MSBuild.exe"
VS_PROJECT_PATH = "project/vs2022/AbccSpiAnalyzer.vcxproj"
RELEASE_TARGET = "-p:Configuration=Release"
DEBUG_TARGET = "-p:Configuration=Debug"
X64_ARCH = "-p:Platform=x64"
X86_ARCH = "-p:Platform=x86"

GNU_CPP_STD = "c++17"

MAC_EXT = ".dylib"
WIN_EXT = ".dll"
LINUX_EXT = ".so"
CPP_EXT = ".cpp"
OBJ_EXT = ".o"

COMPILER = "g++ "
CROSS_COMPILE_32BIT_FLAG = "-m32 "
DYNAMIC_LIB_FLAG = "-dynamiclib "
SHARED_LIB_FLAG = "-shared "

DEBUG_FOLDER = "Debug"

# Specify the search paths/dependencies/options for gcc
INCLUDE_PATHS = ["./sdk/release/include"]
LINK_PATHS = ["./sdk/release/lib"]
INCLUDE_PATHS_DBG = ["./sdk/debug/include"]
LINK_PATHS_DBG = ["./sdk/debug/lib"]

windows_platform = False
mac_platform = False
linux_platform = False
platform_64bit = False


def _build() -> None:
    '''
    Top-level build routine. Builds the plugin for one of the supported host platforms.
    '''

    global mac_platform
    global windows_platform
    global linux_platform
    global platform_64bit

    dylib_ext = ""

    if platform.system().lower() == "windows":
        windows_platform = True
        dylib_ext = WIN_EXT

        if platform.architecture()[0] == "64bit":
            platform_64bit = True

    elif platform.system().lower() == "darwin":
        mac_platform = True
        dylib_ext = MAC_EXT

    else:
        linux_platform = True
        dylib_ext = LINUX_EXT

        if platform.architecture()[0] == "64bit":
            platform_64bit = True

    print("Running on " + platform.system())

    [release_path, release_path32] = _get_release_paths()

    _make_release_dirs(release_path, release_path32)
    _clean_release_dir(release_path, dylib_ext)

    if not mac_platform and platform_64bit:
        _clean_release_dir(release_path32, dylib_ext)

    # The routines called below call exit() and will not return.
    if windows_platform:
        _msbuild()
    else:
        _gnu_build(release_path, release_path32)


def _msbuild() -> None:
    '''
    Routine to compile the plugin using MSBuild.
    '''

    build_error = False

    try:
        # Simply call MSBuild for each pre-configured build target in the
        # pre-configured project.
        if platform_64bit:
            command = [MSBUILD_EXE, VS_PROJECT_PATH, RELEASE_TARGET, X64_ARCH]
            retcode = subprocess.call(command)
            build_error |= (retcode != 0)

            command = [MSBUILD_EXE, VS_PROJECT_PATH, DEBUG_TARGET, X64_ARCH]
            retcode = subprocess.call(command)
            build_error |= (retcode != 0)

        command = [MSBUILD_EXE, VS_PROJECT_PATH, RELEASE_TARGET, X86_ARCH]
        retcode = subprocess.call(command)
        build_error |= (retcode != 0)

        command = [MSBUILD_EXE, VS_PROJECT_PATH, DEBUG_TARGET, X86_ARCH]
        retcode = subprocess.call(command)
        build_error |= (retcode != 0)

    except FileNotFoundError:
        print("ERROR: Build for Windows failed. Ensure MSBuild is part of system's \"path\" variable.")
        print(r'  Example: "C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin"')
        build_error = True

    exit(build_error)


def _gnu_build(release_path: str, release_path32: str = None, dylib_ext: str = LINUX_EXT) -> None:
    '''
    Routine to compile the plugin using G++.

    Parameters
    ----------
    release_path: str
        The output path of the compiled plugin.
    release_path32 : str
        The path where the optional 32-bit plugin will be written to. This
        option shall only be used if on a 64-bit host platform. On a 32-bit
        platform, release_path represents the output path for the plugin.
    dylib_ext: str
        The extension used for the compiled plugin.

    Returns
    ----------
    None
    '''

    build_error = False

    if platform_64bit:
        arch = "64"
    else:
        arch = ""

    debug_path = release_path + DEBUG_FOLDER + "/"

    cpp_files = _get_cpp_file_list()

    if platform_64bit:
        link_dependencies = ["-lAnalyzer" + arch]
        link_dependencies32 = ["-lAnalyzer"]
    else:
        link_dependencies = ["-lAnalyzer"]

    debug_compile_flags = f"-O0 -w -c -fpic -g -std={GNU_CPP_STD} "
    release_compile_flags = f"-O3 -w -c -fpic -std={GNU_CPP_STD} "

    for cpp_file in cpp_files:
        #
        # Generate the command strings for compiling object files
        #
        command = COMPILER
        command_dbg = COMPILER
        obj_file = cpp_file.replace(CPP_EXT, OBJ_EXT)

        for path in INCLUDE_PATHS:
            command += f"-I\"{path}\" "

        for path in INCLUDE_PATHS_DBG:
            command_dbg += f"-I\"{path}\" "

        release_command = command
        release_command += release_compile_flags
        release_command += f"-o \"{release_path}{obj_file}\" "
        release_command += f"\"source/{cpp_file}\""

        if not mac_platform:
            if platform_64bit:
                release_command32 = command
                release_command32 += release_compile_flags
                release_command32 += f"{CROSS_COMPILE_32BIT_FLAG}-o \"{release_path32}{obj_file}\""
                release_command32 += f" \"source/{cpp_file}\""

        debug_command = command_dbg
        debug_command += debug_compile_flags

        if mac_platform:
            debug_command += CROSS_COMPILE_32BIT_FLAG

        debug_command += f"-o \"{debug_path}{obj_file}\" "
        debug_command += f"\"source/{cpp_file}\" -D _DEBUG"

        #
        # Compile object files
        #

        print(release_command)
        retcode = os.system(release_command)
        build_error |= _error_returned(retcode)

        if not mac_platform:
            if platform_64bit:
                print(release_command32)
                retcode = os.system(release_command32)
                build_error |= _error_returned(retcode)

        print(debug_command)
        retcode = os.system(debug_command)
        build_error |= _error_returned(retcode)

    #
    # Generate the command strings for linking object files
    #

    command = COMPILER
    command_dbg = command

    for link_path in LINK_PATHS:
        command += f"-L\"{link_path}\" "

    for link_path in LINK_PATHS_DBG:
        command_dbg += f"-L\"{link_path}\" "

    command32 = command

    for link_dependency in link_dependencies:
        command += link_dependency + " "
        command_dbg += link_dependency + " "

    if platform_64bit:
        for link_dependency in link_dependencies32:
            command32 += link_dependency + " "

    if mac_platform:
        command += DYNAMIC_LIB_FLAG
        command_dbg += DYNAMIC_LIB_FLAG
    else:
        command += SHARED_LIB_FLAG
        command32 += SHARED_LIB_FLAG
        command_dbg += SHARED_LIB_FLAG

    # Figure out what the name of this analyzer is
    analyzer_name = ""

    for cpp_file in cpp_files:
        if cpp_file.endswith("Analyzer.cpp"):
            analyzer_name = cpp_file.replace("Analyzer.cpp", "")
            break

    # The files to create (.so/.dylib files)
    if mac_platform:
        release_command = command + \
            f"-o \"{release_path}lib{analyzer_name}Analyzer{MAC_EXT}\" "
        debug_command = command_dbg + CROSS_COMPILE_32BIT_FLAG + \
            f"-o \"{debug_path}lib{analyzer_name}Analyzer{MAC_EXT}\" "
    else:
        release_command = command + \
            f"-o \"{release_path}lib{analyzer_name}Analyzer{arch}{LINUX_EXT}\" "

        if platform_64bit:
            release_command32 = command32 + CROSS_COMPILE_32BIT_FLAG + \
                f"-o \"{release_path32}lib{analyzer_name}Analyzer{LINUX_EXT}\" "
        debug_command = command_dbg + \
            f"-o \"{debug_path}lib{analyzer_name}Analyzer{arch}{LINUX_EXT}\" "

    # Add all the object files to link
    for cpp_file in cpp_files:
        obj_file = cpp_file.replace(CPP_EXT, OBJ_EXT)
        release_command += f"\"{release_path}{obj_file}\" "

        if platform_64bit:
            release_command32 += f"\"{release_path32}{obj_file}\" "
        debug_command += f"\"{debug_path}{obj_file}\" "

    #
    # Compile dynamic libraries
    #

    print(release_command)
    retcode = os.system(release_command)
    build_error |= _error_returned(retcode)

    if not mac_platform:
        if platform_64bit:
            print(release_command32)
            retcode = os.system(release_command32)
            build_error |= _error_returned(retcode)

        print(debug_command)
        retcode = os.system(debug_command)
        build_error |= _error_returned(retcode)

    else:
        # Only build 32-bit debug library on versions prior to 10.14,
        # newer versions of XCode do not support compiling for 32-bit.
        ver, _, _ = platform.mac_ver()
        ver = float('.'.join(ver.split('.')[:2]))

        if ver < 10.14:
            print(debug_command)
            retcode = os.system(debug_command)
            build_error |= _error_returned(retcode)

    exit(build_error)


def _error_returned(ret_code: int) -> bool:
    '''
    Determines if the return code from a system-call indicates an error.
    '''
    return ret_code != 0


def _get_cpp_file_list() -> list:
    '''
    Returns a list of all the cpp files in /source folder.
    '''
    os.chdir("source")
    cpp_files = glob.glob("*.cpp")
    os.chdir("..")

    return cpp_files


def _get_release_paths() -> list:
    '''
    Returns a list of up to two paths indicating where the compiled output
    artifacts will be written to. One a 32-bit host system there will only be
    one path. On 64-bit systems, with the exception to macOS, there will be
    two paths, one for 32-bit and one for 64-bit. The first element in the
    list will always exist and represent the path that corresponds to the
    native host operating system.
    '''
    release_path_native = None
    release_path32 = None

    if windows_platform:
        if platform_64bit:
            release_path_native = "./plugins/Win64/"
            release_path32 = "./plugins/Win32/"
        else:
            release_path_native = "./plugins/Win32/"

    elif mac_platform:
        release_path_native = "./plugins/OSX/"
    else:
        if platform_64bit:
            release_path_native = "./plugins/Linux64/"
            release_path32 = "./plugins/Linux32/"
        else:
            release_path_native = "./plugins/Linux32/"

    return [release_path_native, release_path32]


def _make_release_dirs(release_path: str, release_path32: str =None) -> None:
    '''
    The specified directory will be created if it does not exist.
    Also, if the platform is compatible with debugging, the associated debug
    path will be created if it does not already exist.

    Parameters
    ----------
    release_path: str
        The output path of the compiled plugin.
    release_path32 : str
        The path where the optional 32-bit plugin will be written to. This
        option shall only be used if on a 64-bit host platform. On a 32-bit
        platform, release_path represents the output path for the plugin.
    dylib_ext: str
        The extension used for the compiled plugin.

    Returns
    ----------
    None
    '''

    if release_path != None and not(windows_platform):
        # NOTE: Unlike other supported targets, the Windows 32-bit debug is
        # sent to the ./debug folder. The visual studio configuration handles
        # this special case.
        release_path += DEBUG_FOLDER

    if release_path32 != None:
        release_path32 += DEBUG_FOLDER

    if release_path != None and not os.path.exists(release_path):
        os.makedirs(release_path)

        if mac_platform:
            print("Made OS plugin directory")
        elif platform_64bit:
            print("Made 64-bit OS plugin directory")
        else:
            print("Made 32-bit OS plugin directory")

    if not mac_platform and release_path32 != None and not os.path.exists(release_path32):
        os.makedirs(release_path32)
        print("Made 32-bit OS plugin directory")


def _clean_release_dir(release_path: str, dylib_ext: str) -> None:
    '''
    Clean the specified folder of output artifacts.

    Parameters
    ----------
    release_path: str
        The output path of the compiled plugin to clean.
    dylib_ext: str
        The extension used for the compiled plugin.

    Returns
    ----------
    None
    '''

    if release_path == None:
        return

    cwd = os.getcwd()
    os.chdir(release_path)

    if windows_platform:
        o_files = glob.glob("./**/*.exp", recursive=True)
        o_files.extend(glob.glob("./**/*.iobj", recursive=True))
        o_files.extend(glob.glob("./**/*.ipdb", recursive=True))
        o_files.extend(glob.glob("./**/*.ilk", recursive=True))
        o_files.extend(glob.glob("./**/*.pdb", recursive=True))
        o_files.extend(glob.glob("./**/*.lib", recursive=True))
    else:
        o_files = glob.glob("./**/*.o", recursive=True)

    o_files.extend(glob.glob("./**/*" + dylib_ext, recursive=True))

    for o_file in o_files:
        os.remove(o_file)

    os.chdir(cwd)


if __name__ == "__main__":
    _build()
